Add validation function

Mxchaeltrxn opened the pull request:
Fixes #34

haydn commented:
Again, I haven’t had a chance to look at this on a computer yet, but from the code it’s look like it’s in the right direction!

A few things I noticed just looking at the code briefly:

- We also need to call the validate function in the Editor where the options are rendered to display the error messages. When I’m at a computer again I can mockup a little visualisation of this so you get what I’m talking about.
- It won’t be necessary to pass the input type to this function. I can’t imagine any cases where a developer would need to change the validation rules based on the input the that’s being used. If they did need to know what the option’s type was for some reason they could work it out from the option’s key (they would have had to define it somewhere else in that config file anyway).

haydn commented:
Here's a example of how/where we want to display the error messages:

![image](https://user-images.githubusercontent.com/27211/73035332-c558b700-3e9b-11ea-8a3e-6727764858fc.png)

Mxchaeltrxn commented:
So I've made the changes you've talked about. 
I added a key argument to validateOptions so that the error message shows above the correct field. What do you think about that?

Also I'm curious as to why (in the issue) you chose the return value to be an object inside an array as opposed to maybe, just an object without the array.

haydn commented on src/Editor.js:
```diff
@@ -1332,99 +1338,106 @@ const Editor = () => {
-                ].options?.map(({ key, input, label }) => {
+                ].options?.map(({ key, input, label }, index) => {
+                  const layer = doc.layers.find(({ id }) => id === [...view.selection][0])
+                  const error = config[layer.component].validate(layer.options, key)
```
We shouldn't need to pass the `key` into the `validate()` function because the data it returns has a `key` property that we can use to get the relevant errors here (the `.?` is an [optional chaining operator](https://github.com/tc39/proposal-optional-chaining) to handle the case where `validate()` returns `null` and the `.find(() => true)` call is used to get the first item off the array):

```suggestion
                  const error = config[layer.component].validate(layer.options)?.filter(error => error.key === key).find(() => true);
```

That means `validate()` functions that check multiple options/rules need to return an array of all errors instead of a single error.

haydn commented on src/Editor.js:
```diff
@@ -1332,99 +1338,106 @@ const Editor = () => {
-                          <Label htmlFor={`option-${key}`}>{label}</Label>
-                          <Textarea
+                        <>
+                          { displayErrorMessage ? <OptionsErrorMessage style={{ paddingTop: index === 0 ? "4px" : "0px", paddingLeft: "6px"}} >{error[0].message}</OptionsErrorMessage> : null }
```
Given the changes above, this can just be:

```suggestion
                          { error ? <OptionsErrorMessage style={{ paddingTop: index === 0 ? "4px" : "0px", paddingLeft: "6px"}} >{error.message}</OptionsErrorMessage> : null }
```

haydn commented on src/Editor.js:
```diff
@@ -1332,99 +1338,106 @@ const Editor = () => {
+                ].options?.map(({ key, input, label }, index) => {
+                  const layer = doc.layers.find(({ id }) => id === [...view.selection][0])
+                  const error = config[layer.component].validate(layer.options, key)
+                  const displayErrorMessage = error && key === error[0].key
```
```suggestion
```

haydn commented on src/config.js:
```diff
@@ -26,7 +27,13 @@ const config = {
+        return [{ key: "text", message: "Must be less than 100 characters long" }];
+      }
+      return null
+    }
```
Without passing `key` this (ad the other like it below) can be:

```js
validate: options =>
  options.text.length > 100
    ? [{ key: "text", message: "Must be less than 100 characters long" }]
    : null
```

Mxchaeltrxn commented on src/config.js:
```diff
@@ -155,6 +156,37 @@ const config = {
       { key: "value", label: "Value", input: "short-string" }
     ]
+  },
+  validateOptions: (options, key) => {
```
I've done this. This makes a lot of sense. Thanks!

Mxchaeltrxn commented:
So I've addressed the problems you've mentioned.

The main reason why I added a key argument to the function was in case an option would have multiple fields with the same name ... but that's impossible. 

Anyway, everything you've said was clear. Thanks!

haydn commented on src/Editor.js:
```diff
@@ -1332,98 +1340,132 @@ const Editor = () => {
+                  );
+                  const error = config[layer.component].validate(
+                    layer.options,
+                    key
```
```suggestion
```

Mxchaeltrxn replied:
I made the change and a few other changes that I must have forgotten to include

haydn commented:
@Mxchaeltrxn Thanks!
